src/mpid/ch4/shm/posix/posix_coll_release_gather.h

```diff
@@ -0,0 +1,276 @@
+    /* Do pipelined release-gather */
+    for (i = 0; i < num_chunks; i++) {
+        int chunk_count = (i == 0) ? chunk_count_floor : chunk_count_ceil;
+
+        mpi_errno =
+            MPIDI_POSIX_mpi_release_gather_release((char *) buffer + offset * extent,
+                                                   chunk_count, datatype, root, comm_ptr,
+                                                   errflag,
+                                                   MPIDI_POSIX_RELEASE_GATHER_OPCODE_BCAST);
+        if (mpi_errno) {
+            /* for communication errors, just record the error but continue */
+            *errflag =
+                MPIX_ERR_PROC_FAILED ==
+                MPIR_ERR_GET_CLASS(mpi_errno) ? MPIR_ERR_PROC_FAILED : MPIR_ERR_OTHER;
+            MPIR_ERR_SET(mpi_errno, *errflag, "**fail");
+            MPIR_ERR_ADD(mpi_errno_ret, mpi_errno);
+        }
+
+        mpi_errno =
+            MPIDI_POSIX_mpi_release_gather_gather(NULL, NULL, 0, MPI_DATATYPE_NULL,
+                                                  MPI_OP_NULL, root, comm_ptr, errflag,
+                                                  MPIDI_POSIX_RELEASE_GATHER_OPCODE_BCAST);
+        if (mpi_errno) {
+            /* for communication errors, just record the error but continue */
+            *errflag =
+                MPIX_ERR_PROC_FAILED ==
+                MPIR_ERR_GET_CLASS(mpi_errno) ? MPIR_ERR_PROC_FAILED : MPIR_ERR_OTHER;
+            MPIR_ERR_SET(mpi_errno, *errflag, "**fail");
+            MPIR_ERR_ADD(mpi_errno_ret, mpi_errno);
+        }
+        offset += chunk_count;
+    }
+
+    if (!is_contig) {
+        if (my_rank != root) {
+            /* Non-root unpack the data if expecting non-contiguous datatypes */
+            position = 0;
+            mpi_errno =
+                MPIR_Unpack_impl(buffer, count, &position, ori_buffer, ori_count, ori_datatype);
+            if (mpi_errno) {
+                /* for communication errors, just record the error but continue */
+                *errflag = MPIR_ERR_OTHER;
+                MPIR_ERR_SET(mpi_errno, *errflag, "**fail");
+                MPIR_ERR_ADD(mpi_errno_ret, mpi_errno);
+            }
+        }
+        MPL_free(buffer);
+    }
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_POSIX_MPI_BCAST_RELEASE_GATHER);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+/* Intra-node reduce is implemented as a release step followed by gather step in release_gather
+ * framework. The actual data movement happens in gather step. Release step makes sure that
+ * the shared reduce buffer can be reused for next reduce call. Release gather framework has
+ * multitple cells in reduce buffer, so that the copying in next cell can be overlapped with
+ * reduction and copying out of previous cells (pipelining).
+ */
+#undef FCNAME
+#define FCNAME MPL_QUOTE(MPIDI_POSIX_mpi_reduce_release_gather)
+MPL_STATIC_INLINE_PREFIX inline int MPIDI_POSIX_mpi_reduce_release_gather(const void *sendbuf,
```

Comment:
In case this slipped your sight, I am rebasing on this code for my PR and I am getting a warning for this duplicated `inline`.

Reply:
I will fix it.